source/logic.rst

```diff
@@ -0,0 +1,492 @@
+    import data.real.basic
+
+    variables (f g : ℝ → ℝ)
+
+    -- BEGIN
+    def even (f : ℝ → ℝ) : Prop := ∀ x, f x = f (-x)
+    def odd (f : ℝ → ℝ) : Prop := ∀ x, f x = - f (-x)
+
+    example (ef : even f) (eg : even g) : even (λ x, f x + g x) :=
+    by { intro x, dsimp, rw [ef, eg] }
+
+    example (of : odd f) (og : odd g) : even (λ x, f x * g x) :=
+    sorry
+
+    example (ef : even f) (og : odd g) : odd (λ x, f x * g x) :=
+    sorry
+
+    example (ef : even f) (og : odd g) : even (λ x, f (g x)) :=
+    sorry
+    -- END
+
+You can find implicit universal quantifiers are all over the place,
+once you know how to spot them.
+Mathlib includes a good library for rudimentary set theory.
+Lean's logical foundation imposes the restriction that when
+we talk about sets, we are always talking about sets of
+elements of some type. If ``x`` has type ``α`` and ``s`` has
+type ``set α``, then ``x ∈ s`` is a proposition that
+asserts that ``x`` is an element of ``s``.
+If ``s`` and ``t`` are of type ``set α``,
+then the subset relation ``s ⊆ t`` is defined to mean
+``∀ {x : α}, x ∈ s → x ∈ t``.
+The variable in the quantifier is marked implicit so that
+given ``h : s ⊆ t`` and ``h' : x ∈ s``,
+we can write ``h h'`` as justification for ``x ∈ t``.
+The following example provides a tactic proof and a proof term
+justifying the reflexivity of the subset relation,
+and asks you to do the same for transitivity.
+
+.. code-block:: lean
+
+    variables {α : Type*} (r s t : set α)
+
+    example : s ⊆ s :=
+    by { intros x xs, exact xs }
+
+    example : s ⊆ s := λ x xs, xs
+
+    example : r ⊆ s → s ⊆ t → r ⊆ t :=
+    begin
+      sorry
+    end
+
+    example : r ⊆ s → s ⊆ t → r ⊆ t :=
+    sorry
+
+Just as we defined ``fn_ub`` for functions,
+we can define ``set_ub s a`` to mean that ``a``
+is an upper bound on the set ``s``,
+assuming ``s`` is a set of elements of some type that
+has an order associated with it.
+In the next example, we ask you to prove that
+if ``a`` is a bound on ``s`` and ``a ≤ b``,
+then ``b`` is a bound on ``s`` as well.
+
+.. code-block:: lean
+
+    variables {α : Type*} [partial_order α]
+    variables (s : set α) (a b : α)
+
+    def set_ub (s : set α) (a : α) := ∀ x, x ∈ s → x ≤ a
+
+    example (h : set_ub s a) (h' : a ≤ b) : set_ub s b :=
+    sorry
+
+We close this section with one last important example.
+A function :math:`f` is said to be *injective* if for
+every :math:`x_1` and :math:`x_2`,
+if :math:`f(x_1) = f(x_2)` then :math:`x_1 = x_2`.
+Mathlib defines ``function.injective f`` with
+``x₁`` and ``x₂`` implicit.
+The next example shows that, on the real numbers,
+any function that adds a constant is injective.
+We then ask you to show that multiplication by a nonzero
+constant is also injective.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    open function
+
+    example (c : ℝ) : injective (λ x, x + c) :=
+    begin
+      intros x₁ x₂ h',
+      apply eq_of_add_eq_add_right h',
```

Comment:
This proof feels like cheating. You could have written the full proof term `λ _ _, eq_of_add_eq_add_right`. And our readers will want to know how to subtract `c` from both side of `h'` and simplify. If we are not ready to explain how to do that then maybe we should choose another example (such that injectivity of a composition of injective functions).

Reply:
I agree, it is not a great example. I wanted readers to look at what happens after the intros. I'll see if I can think of a better one (but maybe I'll leave it since you like the composition one, that comes next).